devolutions-crypto-wayk/src/bastion/fle.rs

```diff
@@ -0,0 +1,203 @@
+    block_size: usize,
+) -> Result<String, Error> {
+    use rand::rngs::OsRng;
+    use rand::Fill;
+
+    let pad_leftover = block_size - (content.len() % block_size);
+    let total_len = AAD_SIZE + IV_SIZE + content.len() + pad_leftover + TAG_SIZE;
+    let mut buffer = vec![0u8; total_len];
+
+    // associated data [magic number (4) | reserved (4) | UUID (16)]
+    let (aad, rest) = buffer.split_at_mut(AAD_SIZE);
+    {
+        let (aad_magic, aad_rest) = aad.split_at_mut(AAD_MAGIC_NUMBER_SIZE);
+        aad_magic.copy_from_slice(&AAD_MAGIC_NUMBER);
+        let (aad_reserved, aad_uuid) = aad_rest.split_at_mut(AAD_RESERVED_SIZE);
+        aad_reserved.copy_from_slice(&[0, 0, 0, 0]); // useless, but just to be extra safe…
+        aad_uuid.copy_from_slice(master_key_id.as_bytes());
+    }
+
+    // nonce / IV
+    let (nonce, rest) = rest.split_at_mut(IV_SIZE);
+    nonce.try_fill(&mut OsRng)?;
+    let nonce = XNonce::from_slice(nonce);
+
+    // actual data
+    let (data, rest) = rest.split_at_mut(rest.len() - TAG_SIZE);
+    data[..content.len()].copy_from_slice(&content);
+    Pkcs7::pad(data, content.len(), block_size).map_err(|_| Error::Padding)?;
+
+    // in-place encryption
+    let key = Key::from_slice(key);
+    let aead = XChaCha20Poly1305::new(key);
+    let tag = aead
+        .encrypt_in_place_detached(&nonce, aad, data)
+        .map_err(|_| Error::XChaCha20)?;
+
+    // finally, write tag
+    rest.copy_from_slice(&tag);
+
+    // our final payload (ie: |~Safe~|k1AjQZYciYvlPmJMnHjbb…)
+    let mut final_payload = String::from(SAFE_PAYLOAD_PREFIX);
+    base64::encode_config_buf(&buffer, base64::STANDARD, &mut final_payload);
+
+    Ok(final_payload)
+}
+
+/// Decrypts field-level content.
+///
+/// - Input is the encrypted key encoded in base64.
+/// - Master Key ID is used for early key mismatch detection.
+/// - Output is decrypted field content.
+///
+/// # Interals
+///
+/// - Decryption is done using XChaCha2020-Poly1035.
+/// - Data layout is assumed to be the same as described in `encrypt_field_with_block_size`.
+pub fn decrypt_field(
+    master_key_id: Uuid,
+    key: &[u8],
+    encrypted_content: &str,
+) -> Result<Vec<u8>, Error> {
+    use chacha20poly1305::aead::Tag;
+
+    const MIN_SIZE: usize = AAD_SIZE + IV_SIZE + 1 + TAG_SIZE;
+
+    let mut buffer = base64::decode_config(
+        &encrypted_content[SAFE_PAYLOAD_PREFIX.len()..],
+        base64::STANDARD,
+    )?;
+
+    if buffer.len() < MIN_SIZE {
+        return Err(Error::InvalidSize { got: buffer.len() });
+    }
+
+    let (aad, rest) = buffer.as_mut_slice().split_at_mut(AAD_SIZE);
+    {
+        // associated data sanity checks
+
+        let (aad_magic, aad_rest) = aad.split_at(AAD_MAGIC_NUMBER_SIZE);
+        if aad_magic != AAD_MAGIC_NUMBER {
+            return Err(Error::InvalidMagicNumber);
+        }
+
+        let (_aad_reserved, uuid) = aad_rest.split_at(AAD_RESERVED_SIZE);
+        let found_master_key_id = Uuid::from_bytes([
```

Comment:
can this be done simply by passing uuid[..16] instead?

Reply:
Good call. Actually, it can't because this is &[u8] and not &[u8; 16] and this is different at the type level.
But I missed [`Uuid::from_slice` ](https://docs.rs/uuid/0.8.2/uuid/struct.Uuid.html#method.from_slice) that could be used instead :facepalm: